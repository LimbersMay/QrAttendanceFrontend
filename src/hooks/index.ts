export * from './useForm';
export * from './useUiStore';
export * from './useAuthStore';
export * from './useCheckAuth';
export * from './useGroupStore';
export * from './useQrCodeStore';
export * from './useRegistryStore';
export * from './useQrAttendanceStore';