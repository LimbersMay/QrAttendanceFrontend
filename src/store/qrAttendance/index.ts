export * from "./group";
export * from "./qrCode";
export * from "./registry";
