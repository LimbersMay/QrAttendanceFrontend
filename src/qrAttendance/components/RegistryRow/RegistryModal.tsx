import React, {ChangeEvent, FormEvent, useEffect, useState} from "react";
import Backdrop from '@mui/material/Backdrop';
import Box from '@mui/material/Box';
import Modal from '@mui/material/Modal';
import Fade from '@mui/material/Fade';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import {Divider, Grid, MenuItem, Select, SelectChangeEvent, TextField} from "@mui/material";
import {useUiSlice} from "../../../hooks/useUiSlice";
import {useRegistrySlice} from "../../../hooks/useRegistrySlice";
import {RegistryDatePicker} from "./RegistryDatePicker";

const style = {
    position: 'absolute' as 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    width: 400,
    bgcolor: 'background.paper',
    border: '2px solid #000',
    boxShadow: 24,
    p: 4,
};

const initialState = {
    name: "",
    checkInTime: "2023-02-18T02:19:01.000Z",
    career: "",
    group: "",
    firstSurname: "",
    secondSurname: "",
}

export const RegistryModal = () => {

    const { active: activeRegistry, updateRegistry } = useRegistrySlice();

    const { isRegistryModalOpen, toggleRegistryModal } = useUiSlice();

    const [formState, setFormState] = useState(activeRegistry || initialState);
    const {name, checkInTime, group, career, firstSurname, secondSurname} = formState;

    useEffect(() => {
        setFormState(activeRegistry || initialState);
    }, [activeRegistry]);

    const changeFormState = ({target}: ChangeEvent<HTMLInputElement> | SelectChangeEvent) => {
        const {name, value} = target;

        setFormState({
            ...formState,
            [name]: value
        })
    }

    const onSelectChange = (event: SelectChangeEvent) => {
        const {name, value} = event.target;

        setFormState({
            ...formState,
            [name]: value
        });
    }

    const onDateChange = (date: string) => {
        setFormState({
            ...formState,
            checkInTime: date
        })
    }

    const onInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        changeFormState(event);
    }

    const onSubmit = (event: FormEvent) => {
        event.preventDefault();

        if (!activeRegistry) return;

        updateRegistry({
            ...activeRegistry,
            ...formState
        });

        console.log(activeRegistry);
        console.log(formState);

        toggleRegistryModal();
    }

    return (
        <Modal
            aria-labelledby="transition-modal-title"
            aria-describedby="transition-modal-description"
            open={isRegistryModalOpen}
            onClose={toggleRegistryModal}
            closeAfterTransition
            BackdropComponent={Backdrop}
            BackdropProps={{
                timeout: 500,
            }}
        >
            <Fade in={isRegistryModalOpen}>
                <Box sx={style}>
                    <form onSubmit={onSubmit}>
                        <Grid container>

                            <Grid item xs={12} sx={{textAlign: "center"}}>
                                <Typography variant="h6">Editing</Typography>
                                <Divider />
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <RegistryDatePicker date={checkInTime} onChangeDate={onDateChange} />
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="Group"
                                    name="group"
                                    value={group}
                                    fullWidth
                                    inputProps={{maxLength: 2, style: {textTransform: 'uppercase'}}}
                                    variant={"outlined"}
                                    placeholder="Group"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <Select fullWidth variant={'outlined'} name="career" value={career} onChange={onSelectChange}>
                                    <MenuItem value="Computer Systems Engineering">Computer Systems Engineering</MenuItem>
                                    <MenuItem value="Civil Engineering">Civil Engineering</MenuItem>
                                    <MenuItem value="Environmental Engineering">Environmental Engineering</MenuItem>
                                    <MenuItem value="Industrial Engineering">Industrial Engineering</MenuItem>
                                    <MenuItem value="Engineering in Administration">Engineering in Administration</MenuItem>
                                    <MenuItem value="Bussines Management Engineering">Bussines Management Engineering</MenuItem>
                                </Select>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="First surname"
                                    name="firstSurname"
                                    value={firstSurname}
                                    fullWidth
                                    variant={"outlined"}
                                    placeholder="firstSurname"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="Second surname"
                                    name="secondSurname"
                                    value={secondSurname}
                                    fullWidth
                                    variant={"outlined"}
                                    placeholder="secondSurname"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <TextField
                                    label="Name"
                                    name="name"
                                    value={name}
                                    fullWidth
                                    variant={"outlined"}
                                    placeholder="name"
                                    onChange={onInputChange}>
                                </TextField>
                            </Grid>

                            <Grid item xs={12} sx={{mt: 2}}>
                                <Button
                                    variant="contained"
                                    fullWidth
                                    type={"submit"}
                                >
                                    Save
                                </Button>
                            </Grid>
                        </Grid>
                    </form>
                </Box>
            </Fade>
        </Modal>
    );
}