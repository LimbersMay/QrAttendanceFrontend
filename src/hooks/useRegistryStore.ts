import {useAppDispatch, useAppSelector} from "../store";
import {
    selectRegistry,
    onSetActiveRegistry,
    updateRegistry, deleteRegistry
} from "../store/qrAttendance";
import {Registry} from "../qrAttendance/interfaces";
import {qrAttendanceApi} from "../api/qrAttendanceApi";

export const useRegistryStore = () => {

    const dispatch = useAppDispatch();
    const {registries, active} = useAppSelector(selectRegistry);

    const startUpdateRegistry = async (registry: Registry) => {
        await qrAttendanceApi.put(`/registry/update`, {
            id: registry.id,
            updatedFields: {
                name: registry.name,
                group: registry.group,
                career: registry.career,
                firstSurname: registry.firstSurname,
                secondSurname: registry.secondSurname,
                checkinTime: registry.checkInTime
            }
        });

        // sync code here
        dispatch(updateRegistry(registry));
    }

    const setActiveRegistry = (registry: Registry | null) => {
        dispatch(onSetActiveRegistry(registry));
    }

    const startDeleteRegistry = async () => {
        const activeRegistryId = `${active?.id}`;

        //  async code here
        await qrAttendanceApi.delete(`/registry/delete/${activeRegistryId}`);

        // sync code here
        dispatch(deleteRegistry(activeRegistryId));
    }

    return {
        // properties
        registries,
        active,

        // methods
        startUpdateRegistry,
        setActiveRegistry,
        startDeleteRegistry
    }
}
