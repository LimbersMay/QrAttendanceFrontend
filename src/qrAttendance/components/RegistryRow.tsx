import {useState} from "react";

import {
    TableCell,
    IconButton,
    TableRow
} from "@mui/material";
import {Delete, Edit, Save} from "@mui/icons-material";

import {useForm} from "../../hooks/useForm";
import {ConditionalTextField} from "./ConditionalTextField";
import {Registry} from "../interfaces";
import {useAppDispatch} from "../../store";
import {startDeleteRegistry, startUpdateRegistry} from "../../store/qrAttendance";

export const RegistryRow = ({registryRow }: { registryRow: Registry}) => {

    const dispatch = useAppDispatch();

    const [isRowEditing, setIsRowEditing] = useState(false);
    const {formState, onInputChange} = useForm(registryRow);

    const { name, sourname, lastname, date } = formState;

    const handleEdit = () => {
        setIsRowEditing(true);
    }

    const handleSave = () => {
        dispatch(startUpdateRegistry({
            ...registryRow,
            ...formState
        }));
        setIsRowEditing(false);
    }

    const handleDelete = () => {
        dispatch(startDeleteRegistry(registryRow.id));
    }

    return (
        <TableRow>
            <TableCell component="th" scope="row">
                <ConditionalTextField
                    name="date"
                    value={date}
                    onChange={onInputChange}
                    condition={isRowEditing}
                />
            </TableCell>
            <TableCell>
                <ConditionalTextField
                    name="name"
                    value={name}
                    onChange={onInputChange}
                    condition={isRowEditing}
                />
            </TableCell>
            <TableCell align="center">
                <ConditionalTextField
                    name="sourname"
                    value={sourname}
                    onChange={onInputChange}
                    condition={isRowEditing}
                    styles={{width: '100px'}}
                />
            </TableCell>
            <TableCell align="center">
                <ConditionalTextField
                    name="lastname"
                    value={lastname}
                    onChange={onInputChange}
                    condition={isRowEditing}
                    styles={{width: '90px'}}
                />
            </TableCell>
            <TableCell align="center">
                {
                    isRowEditing
                        ? <IconButton onClick={handleSave}><Save/></IconButton>
                        : <IconButton onClick={handleEdit}><Edit /></IconButton>
                }
                <IconButton onClick={handleDelete}>
                    <Delete />
                </IconButton>
            </TableCell>
        </TableRow>
    )
}