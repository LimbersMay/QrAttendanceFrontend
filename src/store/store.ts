import {AnyAction, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {authSlice} from "./auth";
import {groupSlice, qrCodeSlice, registrySlice} from "./qrAttendance";
import socketMiddleware from "./middlewares/socketMiddleware";
export const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        group: groupSlice.reducer,
        qrCode: qrCodeSlice.reducer,
        registry: registrySlice.reducer
    },
    middleware: [socketMiddleware]
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    AnyAction
>
