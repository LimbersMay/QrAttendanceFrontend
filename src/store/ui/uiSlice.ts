import {createSlice} from '@reduxjs/toolkit';
import {RootState} from "../store";

interface uiInitialState {
    isQrCodeModalOpen: boolean;
    isRegistryModalOpen: boolean;
    isShowingQrCode: boolean;
    isTitleModalOpen: boolean;
}

const initialState: uiInitialState = {
    isQrCodeModalOpen: false,
    isRegistryModalOpen: false,
    isShowingQrCode: false,
    isTitleModalOpen: false
}

export const uiSlice = createSlice({
    name: 'ui',
    initialState,
    reducers: {
        onShowQrCode: (state) => {
            state.isShowingQrCode = true;
        },
        onHideQrCode: (state) => {
            state.isShowingQrCode = false;
        },
        onToggleQrCodeModal: (state) => {
            state.isQrCodeModalOpen = !state.isQrCodeModalOpen;
        },
        onToggleRegistryModal: (state) => {
            state.isRegistryModalOpen = !state.isRegistryModalOpen;
        },
        onToggleTitleModal: (state) => {
            state.isTitleModalOpen = !state.isTitleModalOpen;
        }
    }
});


// Action creators are generated for each case reducer function
export const { onToggleQrCodeModal, onShowQrCode, onHideQrCode, onToggleRegistryModal, onToggleTitleModal } = uiSlice.actions;
export const selectUi = (state: RootState) => state.ui;