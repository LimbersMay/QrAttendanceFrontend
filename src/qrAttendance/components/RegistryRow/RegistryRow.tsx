import {
    TableCell,
    TableRow
} from "@mui/material";
import {Registry} from "../../interfaces";
import {RegistryRowMenuOptions} from "./RegistryMenuOptions";
import dayjs from "dayjs";
import {RegistryModal} from "./RegistryModal";

export const RegistryRow = ({registryRow}: { registryRow: Registry }) => {

    const {name, group, firstSurname, secondSurname, checkInTime, career} = registryRow;

    return (
        <TableRow>
            <TableCell component="th" scope="row">
                { dayjs(checkInTime).format('MMMM D, YYYY h:mm A') }
            </TableCell>
            <TableCell align="center">
                { group }
            </TableCell>
            <TableCell align="center">
                { career }
            </TableCell>
            <TableCell align="center" sx={{display: {xs: 'none', sm: 'table-cell'}}}>
                { firstSurname }
            </TableCell>
            <TableCell align="center" sx={{display: {xs: 'none', sm: 'table-cell'}}}>
                { secondSurname }
            </TableCell>
            <TableCell align="center">
                { name }
            </TableCell>
            <TableCell align="center" sx={{display: {xs: 'none', sm: 'table-cell'}}}>
                <RegistryRowMenuOptions
                    registry={registryRow}
                />
            </TableCell>

            <RegistryModal />
        </TableRow>
    )
}