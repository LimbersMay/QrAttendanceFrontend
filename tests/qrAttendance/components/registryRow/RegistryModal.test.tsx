import {configureStore} from "@reduxjs/toolkit";
import {uiSlice} from "../../../../src/store/ui/uiSlice";
import {registrySlice} from "../../../../src/store/qrAttendance";
import {withRegistryModalOpenTrue} from "../../../fixtures/uiStates";
import {withActiveRegistryAndRegistriesState} from "../../../fixtures/qrAttendanceStates";
import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import {Provider} from "react-redux";
import {RegistryModal} from "../../../../src/qrAttendance/components/RegistryRow/RegistryModal";
import * as useRegistryStore from "../../../../src/hooks/useRegistryStore";
import * as useUiStore from "../../../../src/hooks/useUiStore";

const store = configureStore({
    reducer: {
        ui: uiSlice.reducer,
        registry: registrySlice.reducer
    },
    preloadedState: {
        ui: {...withRegistryModalOpenTrue},
        registry: {...withActiveRegistryAndRegistriesState}
    }
});

describe('tests for <RegistryModal />', () => {

    beforeEach(() => jest.clearAllMocks());

    test('should match the snapshot', () => {
        const {container} = render(
            <Provider store={store}>
                <RegistryModal />
            </Provider>
        );

        expect(container).toMatchSnapshot();
    });

    test('should show the name of the registry', () => {
        render(
            <Provider store={store}>
                <RegistryModal />
            </Provider>
        );

        const registryName = screen.getByLabelText('registry-title');

        expect(registryName.textContent).toContain(`${withActiveRegistryAndRegistriesState.active?.name}`);
    });

    test('should call the startUpdateRegistry and toggleRegistryModal when submit the form', async () => {

        const mockStartUpdateRegistry = jest.fn();
        const mockToggleRegistryModal = jest.fn();

        const useRegistryStoreSpy = jest.spyOn(useRegistryStore, 'useRegistryStore');
        const useUiStoreSpy = jest.spyOn(useUiStore, 'useUiStore');

        (useRegistryStoreSpy as jest.Mock).mockReturnValue({
            startUpdateRegistry: mockStartUpdateRegistry,
            active: withActiveRegistryAndRegistriesState.active
        });

        (useUiStoreSpy as jest.Mock).mockReturnValue({
            toggleRegistryModal: mockToggleRegistryModal,
            isRegistryModalOpen: true
        });

        render(
            <Provider store={store}>
                <RegistryModal />
            </Provider>
        );

        const registryModalForm = screen.getByLabelText('registry-modal-form');
        fireEvent.submit(registryModalForm);

        await waitFor(() => {
            expect(mockStartUpdateRegistry).toHaveBeenCalled();
            expect(mockToggleRegistryModal).toHaveBeenCalled();
        });
    });
});
