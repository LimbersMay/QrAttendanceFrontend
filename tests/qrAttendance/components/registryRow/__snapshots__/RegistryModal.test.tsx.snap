// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <RegistryModal /> should match the snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
